src/main/scala/io/moia/protos/teleproto/Reader.scala

```diff
@@ -43,18 +43,63 @@ trait Reader[-P, +M] { self =>
     * Transforms successfully read results.
     */
   def map[N](f: M => N): Reader[P, N] =
-    (protobuf: P) => self.read(protobuf).map(f)
+    read(_).map(f)
+
+  /**
+    * Transforms the protobuf before reading.
+    */
+  final def contramap[Q](f: Q => P): Reader[Q, M] =
+    protobuf => read(f(protobuf))
 
   /**
     * Transforms successfully read results with the option to fail.
     */
+  final def emap[N](f: M => PbResult[N]): Reader[P, N] =
```

Comment:
Not sure I like the term emap. I think the design inspiration is coming from circe? 
There you pass in a function to Either - I assume the name is coming from that.
In analogy this should be called `pbmap`?

Reply:
Agreed. I also want to add some tests but didn't find time today.